### PR TITLE
Proposal for RocksDB Compaction optimized for Solana BlockStore

### PR DESCRIPTION
**Pull request by yhchiang-sol**
This document explores RocksDB based solutions for Solana BlockStore
 mentioned in issue [#16234](https://github.com/solana-labs/solana/issues/16234).

### CODE REVIEW

**Comment by yhchiang-sol** [comment]
Please feel free to let me know if any section is unclear or needs more details!

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (steviez) [inline docs/src/proposals/blockstore-rocksdb-compaction.md:164]: What would this look like for writes that are "updates" to metadata fields?

Using the `INDEX_CF` as an example, `INDEX_CF` contains a `BTreeSet<u64>` for received data and coding shreds. Supposing all of our `blockstore::insert_shreds()` calls for this slot have a small number of shreds per insert, we might have 100's (or maybe even > 1000) of updates for a slot. As the slot fills up, the size of the `BTreeSet<u64>` becomes non-trivial:

500 shreds * 8 bytes / shred index * 2 (factoring data and coding) = 8000 bytes (ignoring the overhead of datastructure)

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (yhchiang-sol) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:132]: No.  It will require running manual compaction first (via CompactRange) in order to produce logical levels which level compaction work with.  Otherwise, level compaction will treat every file as L0 file and we will immediately hit write stall,

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (yhchiang-sol) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:85]: We could extend the TTL to be long enough to hold the older ledger.  Btw, can I know more about what is ledger-tool?

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (yhchiang-sol) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:164]: Thanks for pointing this out.  I think I didn't make it clear.

The FIFO Compaction should only be used for the column family that stores the actual shred data.  Metadata stored in different column families should still use level compaction.

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (steviez) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:132]: Got it - thanks!

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (steviez) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:85]: Yep, ledger-tool is a CLI tool that allows manual inspection / manipulation of a ledger. It is in the monorepo at https://github.com/solana-labs/solana/tree/master/ledger-tool, just run `cargo run -- --help` in that sub-folder and you can see the types of commands that it supports.

There are a bunch of read-only operations as well as a few write operations. Obviously the write operations can modify state, but even these write operations are very targeted (ie delete a single slot). In general, I think it be good to have compactions disabled when using the tool.

One caveat that I'll mention is that you can use ledger-tool on a ledger that is actively being updated by a running validator process. In this instance, I think the RocksDB session from the validator would have compaction enabled, and I think this is ok. But, if I just have a ledger on disk sitting around for a week, I wouldn't want simply opening it to print some statistics to cause data to get dropped because it has surpassed a TTL.

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (carllin) [inline docs/src/proposals/blockstore-rocksdb-compaction.md:47]: nit: Delections -> Deletions

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (carllin) [inline docs/src/proposals/blockstore-rocksdb-compaction.md:99]: nit: trigging -> triggering

**Comment by carllin** [comment]
For context we had some discussion about fifo compaction here in the past with @ryoqun https://discord.com/channels/428295358100013066/439194979856809985/804216653918633984

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (carllin) [inline docs/src/proposals/blockstore-rocksdb-compaction.md:163]: nit: diguest

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (carllin) [inline docs/src/proposals/blockstore-rocksdb-compaction.md:105]: For context we had some discussion about fifo compaction here in the past with @ryoqun https://discord.com/channels/428295358100013066/439194979856809985/804218651498643457

I think the primary concern being that the FIFO based column would become out of sync with other columns that are also keyed by slot `X` (inconsistent if one column family contains slot `X` and another doesn't). This is avoided by the current compaction filter which removes from all column families based on the same cutoff at slot `X`

**Comment by carllin** [comment]
Aside from the compaction issues, I would also be interested in benchmarking RocksDb write throughput vs @steviez's simpler file store. 

If I recall correctly, even without compaction, the writes to Rocks were a fair bit slower (writing directly to the memtables). @steviez, @yhchiang-sol and I have already discussed a few ways to approach this (local benchmarks + cluster benchmarks)

**Comment by carllin** [comment]
Also not sure how high priority this issue still is: https://github.com/solana-labs/solana/issues/16234#issuecomment-917071019. Seems like reads from Rocks were causing inserts to stall? @CriesofCarrots do you have more details on this from the RPC side?

If this is a high-pri issue, and we can repro, this might be a better issue to focus on than reworking the compaction.

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (steviez) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:85]: https://github.com/solana-labs/solana/issues/19952 captures the sentiment I was getting at, and apparently our current implementation has shortcomings

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (yhchiang-sol) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:105]: > Before making any compaction changes, I would be interested in benchmarking RocksDb write throughput vs @steviez's simpler file store.
> 
> If I recall correctly, even without compaction, the writes to Rocks were a fair bit slower (writing directly to the memtables). @steviez, @yhchiang-sol and I have already discussed a few ways to approach this (local benchmarks + cluster benchmarks)

Can I know how was the benchmark conducted?  RocksDB by default will run some compactions otherwise the DB will eventually be full.  If we want to benchmark RocksDB w/o compaction, then we need to config it with FIFO.  Or do I miss anything?

**Comment by yhchiang-sol** [comment]
> Also not sure how high priority this issue still is: [#16234 (comment)](https://github.com/solana-labs/solana/issues/16234#issuecomment-917071019). Seems like reads from Rocks were causing inserts to stall? @CriesofCarrots do you have more details on this from the RPC side?
> 
> If this is a high-pri issue, and we can repro, this might be a better issue to focus on than reworking the compaction.

Can I know how was the benchmark conducted? RocksDB by default will run some compactions otherwise the DB will eventually be full. If we want to benchmark RocksDB w/o compaction, then we need to config it with FIFO.

**Comment by yhchiang-sol** [comment]
Some quick updates from the offline discussion with @carllin:

* **Prefer serving fewer shreds over rate-limiting writes to avoid dropping shreds before TTL**:
While dropping shreds before their TTL might be bad when we don't have enough storage to keep all the shreds within the TTL, having validators behind the schedule is a much worse situation.

* **FIFO Compaction first, then transition into manual compaction approach when needed**:
FIFO Compaction offers good amplification factors, but it offers less control on when how shreds exceed the TTL are deleted.  Manual compaction or FIFO compaction with manual compaction will offer the best flexibility while maintaining minimum amplification factors.

* **Benchmarking FIFO Compaction and custom file store**: For benchmarking purposes, comparing FIFO Compaction with custom file store is a good start.

* **Consistency**: FIFO Compaction's TTL actually offers the same level of consistency across multiple column families as the current level compaction approach.  It is because compactions across different column families are not synchronized.  As a result, it is possible that some entries with one slot ID have been dropped by the compaction filter while entries with the same slot ID in other column families are still in the database.

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (yhchiang-sol) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:105]: FIFO compaction's TTL actually offers the same level of consistency across multiple column families as the current level compaction approach we have now. It is because compactions across different column families are not synchronized. As a result, it is possible that some entries with one slot ID have been dropped by the compaction filter while entries with the same slot ID in other column families are still in the database.

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (ryoqun) [inline docs/src/proposals/blockstore-rocksdb-compaction.md:72]: well, when repairing, this might not true. that's why I avoided fifo at the time.

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (ryoqun) [inline docs/src/proposals/blockstore-rocksdb-compaction.md:225]: good point. that's my 2nd concern to switching to non-default fifo, combined with my general lack of familarity with rocksdb (esp, fifo). i think we can now send patches to upstream when using fifo? :)

**Comment by sakridge** [comment]
> Also not sure how high priority this issue still is: [#16234 (comment)](https://github.com/solana-labs/solana/issues/16234#issuecomment-917071019). Seems like reads from Rocks were causing inserts to stall? @CriesofCarrots do you have more details on this from the RPC side?
> 
> If this is a high-pri issue, and we can repro, this might be a better issue to focus on than reworking the compaction.

AFAIK we still have performance problems when doing many blockstore-related RPC queries. That would be interesting to compare vs. this. I did try FIFO compaction back a few months ago when we were having those stalls with the microbenchmark I wrote, but it's possible I didn't enable it in the right way or missed some important configuration option that would have made a difference. I believe that was using the `test_ledger_cleanup_compaction` test.

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (yhchiang-sol) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:225]: Yep, if the benchmark result for FIFO on Solana workload looks promising, I can continue working on all the necessary patches but looks like the RPC side might have higher priority than this.

**Comment by yhchiang-sol** [comment]
> I did try FIFO compaction back a few months ago when we were having those stalls with the microbenchmark I wrote, but it's possible I didn't enable it in the right way or missed some important configuration option that would have made a difference. I believe that was using the test_ledger_cleanup_compaction test.

Thanks for the information!  I tried `test_ledger_cleanup_compaction` and found there is only one writer thread, where it needs multiple writer threads (if I remember it correctly at least 16 writer threads) to fully reach RocksDB's write throughput.

I chatted with @carllin about this, and @carllin mentioned that metadata fields require some synchronization to avoid data race and thus we currently only have one writer thread.  One possible way to relax this synchronization is to use RocksDB's merge operator, which offers read-modify-write without lock by simply appending the delta (for instance, if the operation is +1, then merge operation will simply write "+1" to RocksDB, and those +1s will be "merged" during reads or compactions.

I will take a look at whether using merge operator to enable multiple writers is something possible.

**Comment by ryoqun** [comment]
> @carllin mentioned that metadata fields require some synchronization to avoid data race and thus we currently only have one writer thread.

@yhchiang-sol oh, `merge operator` looks fun to use. also, how about the rocksdb's (pessimistic/optimistic) transactions? a good timing but the rust-rocksdb is about to support it soon (ref: https://github.com/rust-rocksdb/rust-rocksdb/pull/565/files) (btw, i'm one of maintainers of the crate so can prioritize it)

**Comment by yhchiang-sol** [comment]
Created #21657 for the proposal of enabling multiple writers in shred insertion.

**Comment by yhchiang-sol** [comment]
> @yhchiang-sol oh, merge operator looks fun to use. also, how about the rocksdb's (pessimistic/optimistic) transactions? a good timing but the rust-rocksdb is about to support it soon (ref: https://github.com/rust-rocksdb/rust-rocksdb/pull/565/files) (btw, i'm one of maintainers of the crate so can prioritize it).

@ryoqun Using transactions is also an interesting alternative.  I think it's easier to migrate but offers less performance gain compared to that offered by the merge operator.  One interesting aspect is that the two can actually co-exist. So if we have some column families where we have a hard time using the merge operator, we can also consider using both of them.

File: docs/src/proposals/blockstore-rocksdb-compaction.md
Comment (yhchiang-sol) [reply docs/src/proposals/blockstore-rocksdb-compaction.md:85]: > #19952 captures the sentiment I was getting at, and apparently our current implementation has shortcomings
For the ledger tool, we should open the DB with read-only mode.

**Comment by yhchiang-sol** [comment]
I'm updating the proposal that incorporates the feedbacks with the benchmark results.  Will update the PR once it's done!

**Comment by yhchiang-sol** [comment]
Updated the proposal with results from both shred insertion benchmark and mainnet-beta as well.  Remove sections about advanced features to keep the document short.

**Comment by carllin** [comment]
@yhchiang-sol if everything in this proposal is implemented, it might make sense to move this from `proposals/` to `implemented-proposals/`

**Review by carllin — DISMISSED** [review]

**Comment by yhchiang-sol** [comment]
> if everything in this proposal is implemented, it might make sense to move this from proposals/ to implemented-proposals/

The things lefts are the public arguments and migration support.  Will move it to implemented proposals once a the PR for public arguments are merged.

**Pull request merged**
